### ISSUE TITLE
odd lorri loop after a nixos autogc (nix.gc.automatic = true)

### ISSUE THREAD

**Issue by chreekat** [target/lorri#318]
**Describe the bug**

Immediately after an automatic gc, lorri's -vvvv log looks like this, which repeats indefinitely:

```
[2020-01-31T09:33:28Z DEBUG lorri::watch] Watch Event: RawEvent {
    path: Some(
        "/nix/var/nix/profiles/per-user/root"
    ),
    op: Ok(
        CHMOD
    ),
    cookie: None
}
[2020-01-31T09:33:28Z DEBUG lorri::watch] watch event: (
    Ok(
        CHMOD
    ),
    Some(
        "/nix/var/nix/profiles/per-user/root"
    )
)
[2020-01-31T09:33:28Z DEBUG lorri::watch] Received event: RawEvent {
    path: Some(
        "/nix/var/nix/profiles/per-user/root"
    ),
    op: Ok(
        CHMOD
    ),
    cookie: None
}
[2020-01-31T09:33:28Z INFO  lorri::watch] Found 7 events
[2020-01-31T09:33:28Z DEBUG lorri::builder] $ "nix-instantiate" "-vv" "--add-root" "/tmp/.tmpdRr3iQ/result" "--indirect" "--argstr" "runTimeClosure" "/nix/store/1manqjxzqpriw9hq5lfqlc6pgdwfmgbh-runtime-closure.nix" "--argstr" "src" "/home/b/RELEX/repos/isengard/shell.nix" "--" "/home/b/.cache/lorri/cas/4a06de72ed6a4fe178f4a02ad1d642be"

```

The gc was triggered by this NixOS setting:
```nix
...
  nix = {
    gc = {
      automatic = true;
      dates = "11:30";
      options = "--delete-older-than 2w";
    };
  ...
```

**To Reproduce**

1. run `lorri -vvv daemon`
2. wait for the autogc to trigger

**Expected behavior**
No infinite rebuild loop :D

**Metadata**

```console
$ lorri info
lorri version: 1
Lorri Project Configuration

expression: /home/b/RELEX/repos/isengard/shell.nix

```

```console
$ uname -a
Linux bryan-laptop 4.19.93 #1-NixOS SMP Sat Jan 4 18:13:46 UTC 2020 x86_64 GNU/Linux

```

**Additional context**

A fuller log from the time of the GC follows. I see there's also some churn regarding channels.lock, but that part does *not* loop.

```
[2020-01-31T09:30:27Z DEBUG lorri::watch] Event path ("/nix/var/nix/profiles/per-user/root/channels.lock") parent ("/nix/var/nix/profiles/per-user/root") matches watched path
[2020-01-31T09:30:27Z DEBUG lorri::watch] Watch Event: RawEvent {
    path: Some(
        "/nix/var/nix/profiles/per-user/root/channels.lock"
    ),
    op: Ok(
        CREATE
    ),
    cookie: None
}
[2020-01-31T09:30:27Z DEBUG lorri::watch] watch event: (
    Ok(
        CREATE
    ),
    Some(
        "/nix/var/nix/profiles/per-user/root/channels.lock"
    )
)
[2020-01-31T09:30:27Z DEBUG lorri::watch] Event path ("/nix/var/nix/profiles/per-user/root/channels.lock") parent ("/nix/var/nix/profiles/per-user/root") matches watched path
[2020-01-31T09:30:27Z DEBUG lorri::watch] Watch Event: RawEvent {
    path: Some(
        "/nix/var/nix/profiles/per-user/root/channels.lock"
    ),
    op: Ok(
        REMOVE
    ),
    cookie: None
}
[2020-01-31T09:30:27Z INFO  lorri::watch] identified file removal: "/nix/var/nix/profiles/per-user/root/channels.lock"
[2020-01-31T09:30:27Z DEBUG lorri::watch] Received event: RawEvent {
    path: Some(
        "/nix/var/nix/profiles/per-user/root/channels.lock"
    ),
    op: Ok(
        REMOVE
    ),
    cookie: None
}
[2020-01-31T09:30:27Z DEBUG lorri::watch] Event path ("/nix/var/nix/profiles/per-user/root/channels.lock") parent ("/nix/var/nix/profiles/per-user/root") matches watched path
[2020-01-31T09:30:27Z DEBUG lorri::watch] Watch Event: RawEvent {
    path: Some(
        "/nix/var/nix/profiles/per-user/root/channels.lock"
    ),
    op: Ok(
        WRITE
    ),
    cookie: None
}
[2020-01-31T09:30:27Z DEBUG lorri::watch] watch event: (
    Ok(
        WRITE
    ),
    Some(
        "/nix/var/nix/profiles/per-user/root/channels.lock"
    )
)
[2020-01-31T09:30:27Z DEBUG lorri::watch] Received event: RawEvent {
    path: Some(
        "/nix/var/nix/profiles/per-user/root/channels.lock"
    ),
    op: Ok(
        WRITE
    ),
    cookie: None
}
[2020-01-31T09:30:27Z DEBUG lorri::watch] Event path ("/nix/var/nix/profiles/per-user/root/channels.lock") parent ("/nix/var/nix/profiles/per-user/root") matches watched path
[2020-01-31T09:30:27Z DEBUG lorri::watch] Watch Event: RawEvent {
    path: Some(
        "/nix/var/nix/profiles/per-user/root/channels.lock"
    ),
    op: Ok(
        CLOSE_WRITE
    ),
    cookie: None
}
[2020-01-31T09:30:27Z DEBUG lorri::watch] watch event: (
    Ok(
        CLOSE_WRITE
    ),
    Some(
        "/nix/var/nix/profiles/per-user/root/channels.lock"
    )
)
[2020-01-31T09:30:27Z DEBUG lorri::watch] Received event: RawEvent {
    path: Some(
        "/nix/var/nix/profiles/per-user/root/channels.lock"
    ),
    op: Ok(
        CLOSE_WRITE
    ),
    cookie: None
}
[2020-01-31T09:30:27Z INFO  lorri::watch] Found 3 events
Started
[2020-01-31T09:30:27Z DEBUG lorri::builder] $ "nix-instantiate" "-vv" "--add-root" "/tmp/.tmpyhyAwk/result" "--indirect" "--argstr" "runTimeClosure" "/nix/store/1manqjxzqpriw9hq5lfqlc6pgdwfmgbh-runtime-closure.nix" "--argstr" "src" "/home/b/RELEX/repos/isengard/shell.nix" "--" "/home/b/.cache/lorri/cas/4a06de72ed6a4fe178f4a02ad1d642be"
[2020-01-31T09:30:45Z DEBUG lorri::nix] "nix-build" "--out-link" "/tmp/.tmp27T1Jc/result" "--" "/tmp/.tmpyhyAwk/result"
[2020-01-31T09:30:45Z DEBUG lorri::build_loop] original paths: 1922
[2020-01-31T09:30:45Z DEBUG lorri::build_loop]   -> reduced to: 7
[2020-01-31T09:30:45Z DEBUG lorri::watch] Watching parent path "/home/b/RELEX/repos/isengard"
[2020-01-31T09:30:45Z DEBUG lorri::watch] Watching parent path "/home/b/RELEX/repos/isengard"
[2020-01-31T09:30:45Z DEBUG lorri::watch] Watching parent path "/home/b/RELEX/repos/isengard"
[2020-01-31T09:30:45Z DEBUG lorri::watch] Watching parent path "/nix/var/nix/profiles/per-user"
[2020-01-31T09:30:45Z DEBUG lorri::watch] Watching parent path "/home/b/RELEX/repos/isengard/nix"
[2020-01-31T09:30:45Z DEBUG lorri::watch] Watching parent path "/home/b/RELEX/repos/isengard/nix"
[2020-01-31T09:30:45Z DEBUG lorri::watch] Watching parent path "/home/b/.config/nixpkgs"
[2020-01-31T09:30:45Z DEBUG lorri::project::roots] Adding root from "/nix/store/nn3vaxky8a5l711n9wfsc7fw2jzjd8a2-lorri-keep-env-hack-ghc-shell-for-packages" to "/home/b/.cache/lorri/gc_roots/cd222214c9064b2830901f01b5e55b44/gc_root/shell_gc_root"
[2020-01-31T09:30:45Z DEBUG lorri::project::roots] Connecting root from "/home/b/.cache/lorri/gc_roots/cd222214c9064b2830901f01b5e55b44/gc_root/shell_gc_root" to "/nix/var/nix/gcroots/per-user/b/cd222214c9064b2830901f01b5e55b44-shell_gc_root"
Completed(
    BuildResults {
        output_paths: OutputPaths {
            shell_gc_root: RootPath(
                "/home/b/.cache/lorri/gc_roots/cd222214c9064b2830901f01b5e55b44/gc_root/shell_gc_root"
            )
        }
    }
)
[2020-01-31T09:30:45Z DEBUG lorri::watch] Event path ("/nix/var/nix/profiles/per-user/root") directly matches watched path
[2020-01-31T09:30:45Z DEBUG lorri::watch] Watch Event: RawEvent {
    path: Some(
        "/nix/var/nix/profiles/per-user/root"
    ),
    op: Ok(
        CHMOD
    ),
    cookie: None
}
[2020-01-31T09:30:45Z DEBUG lorri::watch] watch event: (
    Ok(
        CHMOD
    ),
    Some(
        "/nix/var/nix/profiles/per-user/root"
    )
)
[2020-01-31T09:30:45Z DEBUG lorri::watch] Event path ("/nix/var/nix/profiles/per-user/root") directly matches watched path
[2020-01-31T09:30:45Z DEBUG lorri::watch] Watch Event: RawEvent {
    path: Some(
        "/nix/var/nix/profiles/per-user/root"
    ),
    op: Ok(
        CHMOD
    ),
    cookie: None
}
[2020-01-31T09:30:45Z DEBUG lorri::watch] watch event: (
    Ok(
        CHMOD
    ),
    Some(
        "/nix/var/nix/profiles/per-user/root"
    )
)
[2020-01-31T09:30:45Z DEBUG lorri::watch] Received event: RawEvent {
    path: Some(
        "/nix/var/nix/profiles/per-user/root"
    ),
    op: Ok(
        CHMOD
    ),
    cookie: None
}
[2020-01-31T09:30:45Z DEBUG lorri::watch] Event path ("/nix/var/nix/profiles/per-user/root") directly matches watched path
[2020-01-31T09:30:45Z DEBUG lorri::watch] Watch Event: RawEvent {
    path: Some(
        "/nix/var/nix/profiles/per-user/root"
    ),
    op: Ok(
        CHMOD
    ),
    cookie: None
}
[2020-01-31T09:30:45Z DEBUG lorri::watch] watch event: (
    Ok(
        CHMOD
    ),
    Some(
        "/nix/var/nix/profiles/per-user/root"
    )
)
[2020-01-31T09:30:45Z DEBUG lorri::watch] Received event: RawEvent {
    path: Some(
        "/nix/var/nix/profiles/per-user/root"
    ),
    op: Ok(
        CHMOD
    ),
    cookie: None
}
[2020-01-31T09:30:45Z DEBUG lorri::watch] Event path ("/nix/var/nix/profiles/per-user/root") directly matches watched path
[2020-01-31T09:30:45Z DEBUG lorri::watch] Watch Event: RawEvent {
    path: Some(
        "/nix/var/nix/profiles/per-user/root"
    ),
    op: Ok(
        CHMOD
    ),
    cookie: None
}
[2020-01-31T09:30:45Z DEBUG lorri::watch] watch event: (
    Ok(
        CHMOD
    ),
    Some(
        "/nix/var/nix/profiles/per-user/root"
    )
)
[2020-01-31T09:30:45Z DEBUG lorri::watch] Received event: RawEvent {
    path: Some(
        "/nix/var/nix/profiles/per-user/root"
    ),
    op: Ok(
        CHMOD
    ),
    cookie: None
}
[2020-01-31T09:30:45Z DEBUG lorri::watch] Event path ("/nix/var/nix/profiles/per-user/root") directly matches watched path
[2020-01-31T09:30:45Z DEBUG lorri::watch] Watch Event: RawEvent {
    path: Some(
        "/nix/var/nix/profiles/per-user/root"
    ),
    op: Ok(
        CHMOD
    ),
    cookie: None
}
[2020-01-31T09:30:45Z DEBUG lorri::watch] watch event: (
    Ok(
        CHMOD
    ),
    Some(
        "/nix/var/nix/profiles/per-user/root"
    )
)
[2020-01-31T09:30:45Z DEBUG lorri::watch] Received event: RawEvent {
    path: Some(
        "/nix/var/nix/profiles/per-user/root"
    ),
    op: Ok(
        CHMOD
    ),
    cookie: None
}
[2020-01-31T09:30:45Z DEBUG lorri::watch] Event path ("/nix/var/nix/profiles/per-user/root") directly matches watched path
[2020-01-31T09:30:45Z DEBUG lorri::watch] Watch Event: RawEvent {
    path: Some(
        "/nix/var/nix/profiles/per-user/root"
    ),
    op: Ok(
        CHMOD
    ),
    cookie: None
}
[2020-01-31T09:30:45Z DEBUG lorri::watch] watch event: (
    Ok(
        CHMOD
    ),
    Some(
        "/nix/var/nix/profiles/per-user/root"
    )
)
[2020-01-31T09:30:45Z DEBUG lorri::watch] Received event: RawEvent {
    path: Some(
        "/nix/var/nix/profiles/per-user/root"
    ),
    op: Ok(
        CHMOD
    ),
    cookie: None
}
[2020-01-31T09:30:45Z DEBUG lorri::watch] Event path ("/nix/var/nix/profiles/per-user/root") directly matches watched path
[2020-01-31T09:30:45Z DEBUG lorri::watch] Watch Event: RawEvent {
    path: Some(
        "/nix/var/nix/profiles/per-user/root"
    ),
    op: Ok(
        CHMOD
    ),
    cookie: None
}
[2020-01-31T09:30:45Z DEBUG lorri::watch] watch event: (
    Ok(
        CHMOD
    ),
    Some(
        "/nix/var/nix/profiles/per-user/root"
    )
)
[2020-01-31T09:30:45Z DEBUG lorri::watch] Received event: RawEvent {
    path: Some(
        "/nix/var/nix/profiles/per-user/root"
    ),
    op: Ok(
        CHMOD
    ),
    cookie: None
}
[2020-01-31T09:30:45Z DEBUG lorri::watch] Event path ("/nix/var/nix/profiles/per-user/root") directly matches watched path
[2020-01-31T09:30:45Z DEBUG lorri::watch] Watch Event: RawEvent {
    path: Some(
        "/nix/var/nix/profiles/per-user/root"
    ),
    op: Ok(
        CHMOD
    ),
    cookie: None
}
[2020-01-31T09:30:45Z DEBUG lorri::watch] watch event: (
    Ok(
        CHMOD
    ),
    Some(
        "/nix/var/nix/profiles/per-user/root"
    )
)
[2020-01-31T09:30:45Z DEBUG lorri::watch] Received event: RawEvent {
    path: Some(
        "/nix/var/nix/profiles/per-user/root"
    ),
    op: Ok(
        CHMOD
    ),
    cookie: None
}
[2020-01-31T09:30:45Z DEBUG lorri::watch] Event path ("/nix/var/nix/profiles/per-user/root") directly matches watched path
[2020-01-31T09:30:45Z DEBUG lorri::watch] Watch Event: RawEvent {
    path: Some(
        "/nix/var/nix/profiles/per-user/root"
    ),
    op: Ok(
        CHMOD
    ),
    cookie: None
}
[2020-01-31T09:30:45Z DEBUG lorri::watch] watch event: (
    Ok(
        CHMOD
    ),
    Some(
        "/nix/var/nix/profiles/per-user/root"
    )
)
[2020-01-31T09:30:45Z DEBUG lorri::watch] Received event: RawEvent {
    path: Some(
        "/nix/var/nix/profiles/per-user/root"
    ),
    op: Ok(
        CHMOD
    ),
    cookie: None
}
[2020-01-31T09:30:45Z DEBUG lorri::watch] Event path ("/nix/var/nix/profiles/per-user/root") directly matches watched path
[2020-01-31T09:30:45Z DEBUG lorri::watch] Watch Event: RawEvent {
    path: Some(
        "/nix/var/nix/profiles/per-user/root"
    ),
    op: Ok(
        CHMOD
    ),
    cookie: None
}
[2020-01-31T09:30:45Z DEBUG lorri::watch] watch event: (
    Ok(
        CHMOD
    ),
    Some(
        "/nix/var/nix/profiles/per-user/root"
    )
)
[2020-01-31T09:30:45Z DEBUG lorri::watch] Received event: RawEvent {
    path: Some(
        "/nix/var/nix/profiles/per-user/root"
    ),
    op: Ok(
        CHMOD
    ),
    cookie: None
}
[2020-01-31T09:30:45Z DEBUG lorri::watch] Event path ("/nix/var/nix/profiles/per-user/root") directly matches watched path
[2020-01-31T09:30:45Z DEBUG lorri::watch] Watch Event: RawEvent {
    path: Some(
        "/nix/var/nix/profiles/per-user/root"
    ),
    op: Ok(
        CHMOD
    ),
    cookie: None
}
[2020-01-31T09:30:45Z DEBUG lorri::watch] watch event: (
    Ok(
        CHMOD
    ),
    Some(
        "/nix/var/nix/profiles/per-user/root"
    )
)
[2020-01-31T09:30:45Z DEBUG lorri::watch] Received event: RawEvent {
    path: Some(
        "/nix/var/nix/profiles/per-user/root"
    ),
    op: Ok(
        CHMOD
    ),
    cookie: None
}
[2020-01-31T09:30:45Z DEBUG lorri::watch] Event path ("/nix/var/nix/profiles/per-user/root") directly matches watched path
[2020-01-31T09:30:45Z DEBUG lorri::watch] Watch Event: RawEvent {
    path: Some(
        "/nix/var/nix/profiles/per-user/root"
    ),
    op: Ok(
        CHMOD
    ),
    cookie: None
}
[2020-01-31T09:30:45Z DEBUG lorri::watch] watch event: (
    Ok(
        CHMOD
    ),
    Some(
        "/nix/var/nix/profiles/per-user/root"
    )
)
[2020-01-31T09:30:45Z DEBUG lorri::watch] Received event: RawEvent {
    path: Some(
        "/nix/var/nix/profiles/per-user/root"
    ),
    op: Ok(
        CHMOD
    ),
    cookie: None
}
[2020-01-31T09:30:45Z INFO  lorri::watch] Found 11 events
Started
[2020-01-31T09:30:45Z DEBUG lorri::builder] $ "nix-instantiate" "-vv" "--add-root" "/tmp/.tmp5vW46F/result" "--indirect" "--argstr" "runTimeClosure" "/nix/store/1manqjxzqpriw9hq5lfqlc6pgdwfmgbh-runtime-closure.nix" "--argstr" "src" "/home/b/RELEX/repos/isengard/shell.nix" "--" "/home/b/.cache/lorri/cas/4a06de72ed6a4fe178f4a02ad1d642be"
[2020-01-31T09:30:49Z DEBUG lorri::nix] "nix-build" "--out-link" "/tmp/.tmpOfh3fq/result" "--" "/tmp/.tmp5vW46F/result"
[2020-01-31T09:30:49Z DEBUG lorri::build_loop] original paths: 1922
[2020-01-31T09:30:49Z DEBUG lorri::build_loop]   -> reduced to: 7
[2020-01-31T09:30:49Z DEBUG lorri::watch] Watching parent path "/nix/var/nix/profiles/per-user"
[2020-01-31T09:30:49Z DEBUG lorri::watch] Watching parent path "/home/b/RELEX/repos/isengard"
[2020-01-31T09:30:49Z DEBUG lorri::watch] Watching parent path "/home/b/RELEX/repos/isengard/nix"
[2020-01-31T09:30:49Z DEBUG lorri::watch] Watching parent path "/home/b/RELEX/repos/isengard"
[2020-01-31T09:30:49Z DEBUG lorri::watch] Watching parent path "/home/b/RELEX/repos/isengard/nix"
[2020-01-31T09:30:49Z DEBUG lorri::watch] Watching parent path "/home/b/RELEX/repos/isengard"
[2020-01-31T09:30:49Z DEBUG lorri::watch] Watching parent path "/home/b/.config/nixpkgs"
[2020-01-31T09:30:49Z DEBUG lorri::project::roots] Adding root from "/nix/store/nn3vaxky8a5l711n9wfsc7fw2jzjd8a2-lorri-keep-env-hack-ghc-shell-for-packages" to "/home/b/.cache/lorri/gc_roots/cd222214c9064b2830901f01b5e55b44/gc_root/shell_gc_root"
[2020-01-31T09:30:49Z DEBUG lorri::project::roots] Connecting root from "/home/b/.cache/lorri/gc_roots/cd222214c9064b2830901f01b5e55b44/gc_root/shell_gc_root" to "/nix/var/nix/gcroots/per-user/b/cd222214c9064b2830901f01b5e55b44-shell_gc_root"
Completed(
    BuildResults {
        output_paths: OutputPaths {
            shell_gc_root: RootPath(
                "/home/b/.cache/lorri/gc_roots/cd222214c9064b2830901f01b5e55b44/gc_root/shell_gc_root"
            )
        }
    }
)
[2020-01-31T09:30:49Z DEBUG lorri::watch] Event path ("/nix/var/nix/profiles/per-user/root") directly matches watched path
[2020-01-31T09:30:49Z DEBUG lorri::watch] Watch Event: RawEvent {
    path: Some(
        "/nix/var/nix/profiles/per-user/root"
    ),
    op: Ok(
        CHMOD
    ),
    cookie: None
}
```

**Comment by curiousleo** [comment]
Thanks for this report!

This looks like the kind of issue fixed by https://github.com/target/lorri/pull/260. We have released lorri to `nixos-unstable` and `nixos-19.09` since the bug was fixed: https://github.com/NixOS/nixpkgs/pull/77380. Have you updated your channels recently?

**Comment by chreekat** [comment]
Oh! I was still using the rolling release, pulling lorri directly from this repo :see_no_evil: I'll move to using it through nixos-19.09 directly, and then I should be good to go! In fact, feel free to close this now. I'll just re-open if I experience it again.
